controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,650 @@
+	Watch(ctx context.Context, input WatchInput) error
+
+	// GetLastProbeSuccessTimestamp returns the time when the health probe was successfully executed last.
+	GetLastProbeSuccessTimestamp(ctx context.Context, cluster client.ObjectKey) time.Time
+
+	// GetClusterSource returns a Source of Cluster events.
+	// GenericEvents are always sent on connect and disconnect.
+	// Additionally the WatchForProbeFailure(time.Duration) option can be used to get an event if the health probe
+	// didn't succeed for the configured duration.
+	// Note: GetClusterSource would ideally take a mapFunc that has a *clusterv1.Cluster instead of a client.Object
+	// as a parameter, but then our existing mapFuncs wouldn't work and we would have to implement new ones.
+	GetClusterSource(mapFunc func(ctx context.Context, cluster client.Object) []ctrl.Request, opts ...GetClusterSourceOption) source.Source
+}
+
+// ErrClusterNotConnected is returned by the ClusterCache when e.g. a Client cannot be returned
+// because there is no connection to the workload cluster.
+var ErrClusterNotConnected = errors.New("connection to the workload cluster is down")
+
+// Watcher is a scoped-down interface from Controller that only knows how to watch.
+type Watcher interface {
+	// Watch watches the provided Source.
+	Watch(src source.Source) error
+}
+
+// WatchInput specifies the parameters used to establish a new watch for a workload cluster.
+// A source.Kind source (configured with Kind, EventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.Kind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+type WatchInput struct {
+	// Name represents a unique Watch request for the specified Cluster.
+	// The name is used to track that a specific watch is only added once to a cache.
+	// After a connection (and thus also the cache) has been re-created, watches have to be added
+	// again by calling the Watch method again.
+	Name string
+
+	// Cluster is the object key of the workload cluster.
+	Cluster client.ObjectKey
+
+	// Watcher is the watcher (controller) whose Reconcile() function will be called for events.
+	Watcher Watcher
+
+	// Kind is the type of resource to watch.
+	Kind client.Object
+
+	// EventHandler contains the event handlers to invoke for resource events.
+	EventHandler handler.EventHandler
+
+	// Predicates is used to filter resource events.
+	Predicates []predicate.Predicate
+}
+
+// GetClusterSourceOption is an option that modifies GetClusterSourceOptions for a GetClusterSource call.
+type GetClusterSourceOption interface {
+	// ApplyToGetClusterSourceOptions applies this option to the given GetClusterSourceOptions.
+	ApplyToGetClusterSourceOptions(option *GetClusterSourceOptions)
+}
+
+// GetClusterSourceOptions allows to set options for the GetClusterSource method.
+type GetClusterSourceOptions struct {
+	watchForProbeFailures []time.Duration
```

Comment:
q: why this should be an array?

Reply:
To allow one controller to be notified not just after one duration.

Example, Cluster controller wants to be notified:
* after --remote-connection-grace-period to update the RemoteConnectionProbe
* after --remote-conditions-grace-period to update remote conditions

While we might not have exactly that use case, I think it's not an unrealistic one. So I would like to keep that functionality.